### PR TITLE
nvme/plugins: Add performance stats v2

### PR DESCRIPTION
**Pull request by jinhuahuang2021**
Add performance stats v2

### CODE REVIEW

File: plugins/memblaze/memblaze-nvme.c [hunk 1]
```diff
@@ -107,7 +107,8 @@ static __u64 raw_2_u64(const __u8 *buf, size_t len)
 	return le64_to_cpu(val);
 }
 
-static void get_memblaze_new_smart_info(struct nvme_p4_smart_log *smart, int index, __u8 *nm_val, __u8 *raw_val)
+static void get_memblaze_new_smart_info(struct nvme_p4_smart_log *smart, int index, __u8 *nm_val,
+										__u8 *raw_val)
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:111]: I'd prefer if you could split out white space changes into a separate patch. Mixing janitor updates and new code is making the review harder that it needs to be.

Anyway, when shortening overlong lines, please align the arguments or use two spaces as indent, e.g.

```c
static void get_memblaze_new_smart_info(struct nvme_p4_smart_log *smart, int index, __u8 *nm_val,
					__u8 *raw_val)
```

or 

```c
static void get_memblaze_new_smart_info(struct nvme_p4_smart_log *smart, int index, __u8 *nm_val,
		__u8 *raw_val)
```

File: plugins/memblaze/memblaze-nvme.c [hunk 2]
```diff
@@ -144,35 +145,42 @@ static void show_memblaze_smart_log_new(struct nvme_memblaze_smart_log *s,
 		"min: ", *(__u16 *)raw, ", max: ", *(__u16 *)(raw+2), ", avg: ", *(__u16 *)(raw+4));
 
 	get_memblaze_new_smart_info(smart, RAISIN_SI_VD_E2E_DECTECTION_COUNT, nm, raw);
-	printf("%-31s: %3d%%       %"PRIu64"\n", "end_to_end_error_detection_count", *nm, int48_to_long(raw));
+	printf("%-31s: %3d%%       %"PRIu64"\n", "end_to_end_error_detection_count", *nm,
+			int48_to_long(raw));
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:149]: Here the `int48_to_long` argument could align with the opening '('.

Obviously, this is also true for all cases below.

File: plugins/memblaze/memblaze-nvme.c [hunk 3]
```diff
@@ -144,35 +145,42 @@ static void show_memblaze_smart_log_new(struct nvme_memblaze_smart_log *s,
...
 
 	get_memblaze_new_smart_info(smart, RAISIN_SI_VD_TIMED_WORKLOAD_TIMER, nm, raw);
-	printf("%-32s: %3d%%       %"PRIu64"%s\n", "timed_workload_timer", *nm, int48_to_long(raw), " min");
+	printf("%-32s: %3d%%       %"PRIu64"%s\n", "timed_workload_timer", *nm, int48_to_long(raw),
+			" min");
 
 	get_memblaze_new_smart_info(smart, RAISIN_SI_VD_THERMAL_THROTTLE_STATUS, nm, raw);
 	printf("%-32s: %3d%%       %u%%%s%"PRIu64"\n", "thermal_throttle_status", *nm,
-	       *raw, ", cnt: ", int48_to_long(raw+1));
+		   *raw, ", cnt: ", int48_to_long(raw+1));
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:168]: This change is not needed as the line is already properly align.

File: plugins/memblaze/memblaze-nvme.c [hunk 4]
```diff
@@ -402,11 +414,12 @@ int parse_params(char *str, int number, ...)
 	return 0;
 }
 
-static int mb_get_additional_smart_log(int argc, char **argv, struct command *cmd, struct plugin *plugin)
+static int mb_get_additional_smart_log(
+	int argc, char **argv,
+	struct command *cmd, struct plugin *plugin)
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:419]: Indent these two lines with double tap, so the arguments do not align with the single tab indented variable declearations.

File: plugins/memblaze/memblaze-nvme.c [hunk 5]
```diff
@@ -402,11 +414,12 @@ int parse_params(char *str, int number, ...)
...
 {
 	struct nvme_memblaze_smart_log smart_log;
-	char *desc =
-	    "Get Memblaze vendor specific additional smart log (optionally, for the specified namespace), and show it.";
+	char *desc = "Get Memblaze additional smart log, and show it.";
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:422]: Overlong strings are okay to be one line. This is an exception but if the shortened help text is also fine. Either way you like it.

File: plugins/memblaze/memblaze-nvme.c [hunk 6]
```diff
@@ -435,7 +448,7 @@ static int mb_get_additional_smart_log(int argc, char **argv, struct command *cm
 	if (!err) {
 		if (!cfg.raw_binary)
 			err = show_memblaze_smart_log(dev_fd(dev), cfg.namespace_id, dev->name,
-						      &smart_log);
+							  &smart_log);
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:451]: Not needed.

File: plugins/memblaze/memblaze-nvme.c [hunk 7]
```diff
@@ -563,7 +582,9 @@ static int mb_set_powermanager_status(int argc, char **argv, struct command *cmd
 #define P2MIN											   (1)
 #define P2MAX											   (5000)
 #define MB_FEAT_HIGH_LATENCY_VALUE_SHIFT				   (15)
-static int mb_set_high_latency_log(int argc, char **argv, struct command *cmd, struct plugin *plugin)
+static int mb_set_high_latency_log(int argc, char **argv,
+									struct command *cmd,
+									struct plugin *plugin)
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:587]: Too many tab indention for the arguments.

File: plugins/memblaze/memblaze-nvme.c [hunk 8]
```diff
@@ -667,7 +688,7 @@ struct log_page_high_latency {
 static int find_deadbeef(char *buf)
 {
 	if (((*(buf + 0) & 0xff) == 0xef) && ((*(buf + 1) & 0xff) == 0xbe) &&
-	    ((*(buf + 2) & 0xff) == 0xad) && ((*(buf + 3) & 0xff) == 0xde))
+		((*(buf + 2) & 0xff) == 0xad) && ((*(buf + 3) & 0xff) == 0xde))
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:691]: Not needed

File: plugins/memblaze/memblaze-nvme.c [hunk 9]
```diff
@@ -1840,3 +1900,529 @@ static int mb_get_latency_feature(int argc, char **argv, struct command *cmd,
...
+			d_raw((unsigned char *)&log, sizeof(struct latency_stats));
+	} else if (err > 0) {
+		nvme_show_status(err);
+	} else {
+		nvme_show_error("%s: %s", cmd->name, nvme_strerror(errno));
+	}
+
+	// Close device
+
+	dev_close(dev);
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:2051]: You could use the cleanup helper and then you could just remove the `dev_close`.

```c
        _cleanup_nvme_dev_ struct nvme_dev *dev = NULL;
```

File: plugins/memblaze/memblaze-nvme.c [hunk 10]
```diff
@@ -1840,3 +1900,529 @@ static int mb_get_latency_feature(int argc, char **argv, struct command *cmd,
...
+		fprintf(stderr, "Major Version %u: Not supported yet\n", major_version);
+		break;
+	}
+}
+
+static int mb_get_latency_stats(int argc, char **argv,
+								struct command *cmd,
+								struct plugin *plugin)
+{
+	// Get the configuration
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:2008]: Please avoid non meaningful comments. It's obvious from the code what is happening. This clutters just the code. Some for the rest of the comments in this function.

File: plugins/memblaze/memblaze-nvme.c [hunk 11]
```diff
@@ -1840,3 +1900,529 @@ static int mb_get_latency_feature(int argc, char **argv, struct command *cmd,
...
+
+	default:
+		fprintf(stderr, "Major Version %u: Not supported yet\n", major_version);
+		break;
+	}
+}
+
+static int mb_get_latency_stats(int argc, char **argv,
+								struct command *cmd,
+								struct plugin *plugin)
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:2006]: Indention is off.

File: plugins/memblaze/memblaze-nvme.c [hunk 12]
```diff
@@ -1840,3 +1900,529 @@ static int mb_get_latency_feature(int argc, char **argv, struct command *cmd,
...
+		char   str_time_s[20]  = {0};
+		char   str_time_ms[32] = {0};
+
+		strftime(str_time_s, sizeof(str_time_s),
+			"%Y-%m-%d %H:%M:%S", localtime(&timestamp_s));
+		snprintf(str_time_ms, sizeof(str_time_ms),
+			"%s.%03d", str_time_s, time_ms);
+		printf("%-24s", str_time_ms);
+
+		//
```
Comment (igaw) [inline plugins/memblaze/memblaze-nvme.c:2080]: remove this empty comment

**Review by igaw — CHANGES_REQUESTED** [review]
some nitpicking:
- split the white space changes from the new code 
- pay more attention on proper alignment
- use the new cleanup helpers for `struct nvme_dev` allocations

Rest looks okay.

**Comment by jinhuahuang2021** [comment]
The PR need more change, I will update it together.

**Comment by igaw** [comment]
Do you still work on this?

**Comment by jinhuahuang2021** [comment]
Yes, we are :)

**Comment by igaw** [comment]
Okay, not sure what the update from yesterday means. FWIW, there shouldn't be any merges in the PR. Just rebase your changes to the latest upstream version and update the commit from there. Hope this helps.

**Comment by jinhuahuang2021** [comment]
This PR contains various modifications, which makes it confusing.
I close this PR and split it to small PRs.

**Comment by jinhuahuang2021** [comment]
This PR contains various modifications, which makes it confusing.
I close this PR and split it to small PRs.

**Pull request closed without merging**